File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.HashSet;
import java.util.Set;
class Solution {
    
    public int[] parent;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        parent = new int[n];
        for(int i = 0 ; i < n ; i++){
            parent[i] = i;
        }
        
        for(int i = 0 ; i < n ; i++){
            for(int j = 0 ; j < n ; j++){
                if(i == j || computers[i][j] == 0) continue;
                
                if(find(i) != find(j)){
                    union(i,j);
                }
            }
        }
        Set<Integer> count = new HashSet<>();
        for(int i = 0 ; i < n ; i++){
            // System.out.println(parent[i]);
            count.add(find(parent[i]));
        }
        return count.size();
    }
    
    public void union(int a, int b){
        int pa = find(a);
        int pb = find(b);
        parent[Math.max(pa, pb)] = Math.min(pa,pb);
    }
    
    public int find(int a){
        if(parent[a] != a){
            parent[a] = find(parent[a]);
        }
        return parent[a];
    }
    
     
}

File: 프로그래머스/3/43162. 네트워크/네트워크.kt
class Solution {
    fun solution(n: Int, computers: Array<IntArray>): Int {
        var parent = (0 until n).toMutableList()
        for(i in 0 until n){
            for(j in 0 until n){
                if(i == j || computers[i][j] == 0){
                    continue
                }
                val na = find(parent, i)
                val nb = find(parent, j)
                if(na < nb){
                    parent[nb] = na
                } else if (na > nb){
                    parent[na] = nb
                }
            }
        }
        for(i in 0 until n){
            parent[i] = find(parent, i)
        }
        return parent.toSet().size
    }

    fun find(parent: List<Int>, a: Int) :Int{
        if(parent[a] == a){
           return a
        }
        return find(parent, parent[a])
    }
}